context/src/otelInGrpcTest/java/io/opentelemetry/context/GrpcContextStorageProvider.java

```diff
@@ -27,15 +26,58 @@ public Scope attach(Context toAttach) {
         return Scope.noop();
       }
 
-      io.grpc.Context newGrpcContext = grpcContext.withValue(OTEL_CONTEXT, toAttach);
-      io.grpc.Context toRestore = newGrpcContext.attach();
+      io.grpc.Context newGrpcContext;
+      if (toAttach instanceof GrpcContextWrapper) {
+        // This was already constructed with an embedded grpc Context.
+        newGrpcContext = ((GrpcContextWrapper) toAttach).toGrpcContext();
+      } else {
+        newGrpcContext = grpcContext.withValue(OTEL_CONTEXT, toAttach);
+      }
 
+      io.grpc.Context toRestore = newGrpcContext.attach();
       return () -> newGrpcContext.detach(toRestore);
     }
 
     @Override
     public Context current() {
-      return OTEL_CONTEXT.get();
+      // We return an object that embeds both the
+      io.grpc.Context grpcContext = io.grpc.Context.current();
```

Comment:
From what I understand, we essentially resync gRPC with OTel context any time `current` is called. And since we naturally can't have implicit context propagation without calling `current` anyways this covers all use cases. Does that sound right?

Reply:
Yes. This is similar to otelAsGRPC but instead of getting values from gRPC context using the wrapped keys, you first get our context from there, then use the key to get the value from our context.

We can simplify the code to always call into the gRPC via the extra call I mentioned but that is a bit more inefficient.